src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1346,23 +1358,30 @@ internal void FinishWrite()
 
             lock (this)
             {
-                HandleWriteCallback();
+                HandleQueuedCallback(ref _queuedWriteStateRequest);
             }
         }
 
-        private void HandleWriteCallback()
+        private void HandleQueuedCallback(ref object queuedStateRequest)
         {
-            object obj = _queuedWriteStateRequest;
-            _queuedWriteStateRequest = null;
+            object obj = queuedStateRequest;
+            if (obj == null)
+            {
+                return;
+            }
+            queuedStateRequest = null;
+                        
             switch (obj)
             {
-                case null:
-                    break;
                 case LazyAsyncResult lazy:
                     lazy.InvokeCallback();
                     break;
-                case TaskCompletionSource<int> tsc:
-                    tsc.SetResult(0);
+                case TaskCompletionSource<int> taskCompletionSource when taskCompletionSource.Task.AsyncState != null:
+                    Memory<byte> array = (Memory<byte>)taskCompletionSource.Task.AsyncState;
+                    taskCompletionSource.SetResult(CheckOldKeyDecryptedData(array));
+                    break;
```

Comment:
How common is this case?  If it's very common, and if we expect it to be common that the `Memory<byte>` will be wrapping an entire array rather than something other than an array or part of an array, for a subsequent change we might want to look at adding an "fast path" where the AsyncState wraps the array rather than wrapping the `Memory<byte>`.  Then here we'd just reconstitute the `Memory<byte>` from it, e.g.
```C#
case TaskCompletionSource<int> taskCompletionSource when taskCompletionSource.Task.AsyncState != null:
    if (taskCompletionSource.Task.AsyncState is byte[] array)
    {
        taskCompletionSource.SetResult(CheckOldKeyDecryptedData(array));
    }
    else
    {
        taskCompletionSource.SetResult(CheckOldKeyDecryptedData((Memory<byte>)taskCompletionSource.Task.AsyncState);
    }
    break;
```
That would save a boxing allocation for such cases, but at the expense of a type check here, so it'd only be worthwhile if it'd likely be a common case.